### DISCUSSION TITLE
Can we abandon html-janitor and change to anything else (i.e. sanitize-html)

### DISCUSSION THREAD

**Discussion by radityoardi** [codex-team/editor.js#2184] [Ideas]
## Background
I've been spending hours looking at the code while writing a custom Tool that can process the following tag (usually for highlightjs).
```
<pre>
    <code class="powershell">
        #this is comment for powershell
        .\AndThisIsTheScript.ps1 -Calling
    </code>
</pre>
```
And I've been using this piece of code to populate editorjs with HTML input.
```
editor.blocks.renderFromHTML(htmlString);
```
But because of I want to let the end-users choose the language they want using a dropdown, I had to customize the code plugin at https://github.com/editor-js/code. Just so that the code looks like this:
![image](https://user-images.githubusercontent.com/12979499/204076763-be32c811-6774-4ece-a3fe-33fd20b4a39c.png)

## Issue
It was amazing actually, parsing data was okay, everything works perfectly as expected. Except, digging further deep down, it seems like I can't easily parse HTML so that the drop down reads correctly from the HTML. No matter what HTML I put in, the CODE tag was stripped out including its attributes. I found out that the code was using a component to sanitize the HTML tags and attributes, relying on https://github.com/guardian/html-janitor.

And guess what? The code maintainer has archived this code and wasn't updated for years.

### What I have tried
Play around with sanitize and onPaste method. This code `return { code: true }` doesn't work, `return { code: { 'class': true } }` not working also.

## Idea
Instead of relying on this old component, why don't we use a component that's actively maintained like https://www.npmjs.com/package/sanitize-html? Or any other active component (you suggest)?
I do believe there will be a major breaking change especially dealing with plugins created by amazing people here separately.

**Comment by radityoardi** [comment]
Just an additional note, I also notice that there is current standing issue related to TABLE seen at the comment of the code below, which is pretty similar to what I've been experiencing right now.
https://github.com/codex-team/editor.js/blob/a8026e68296dd9216bc94f4dee51566aff44182e/src/components/modules/paste.ts#L563

And to mention some other paste issues like list items, or other complex html structures. Well, the idea is NOT to entertain those complex html structures, but to avoid with unnecessary fixes while the html-janitor is actually not actively maintained.

**Comment by radityoardi** [comment]
let me see what I can do, I'm forking out the next branch and working it on my local first.
https://github.com/radityoardi/editor.js/tree/html-janitor-replacement

**Reply by ARH3** [reply]
Hey @radityoardi - Any luck with this? Do you intend to send a pull request with the new janitor once you figure it out? See my note below for the related issue I am experiencing.

**Reply by ARH3** [reply]
Hey @radityoardi just checking in one more time. Have you been able to look into this?

**Comment by ARH3** [comment]
Adding my support to move to a different janitor! I am having a different problem also caused by html-janitor (I believe)

onPaste, I want to keep "&nbsp;" even when it is in between elements, ie: <span>some text</span>&nbsp;&nbsp;<span>some more text</span>.

I've noticed this works fine when I am pasting within a block that I specify in pasteConfig... but if it's an inline paste... it gets stripped every time. I am 90% sure it is due to this part of the html-janitor:
<img width="787" alt="image" src="https://user-images.githubusercontent.com/2137208/210273098-354a9d2e-8042-4637-adc8-f0adcc0e8764.png">

Source: https://github.com/guardian/html-janitor/blob/master/src/html-janitor.js

Any advice or workarounds appreciated... but otherwise I am hoping that if you switch to a new janitor, it removing white space like this becomes something we can specify. Thanks!!
